cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/FileBatchSource.java

```diff
@@ -0,0 +1,201 @@
+      pipelineConfigurer.createDataset(config.timeTable, KeyValueTable.class, DatasetProperties.EMPTY);
+    }
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    //calculate date one minute ago, rounded down to the nearest minute
+    prevMinute = new Date(context.getLogicalStartTime() - TimeUnit.MINUTES.toMillis(1));
+    Calendar cal = Calendar.getInstance();
+    cal.setTime(prevMinute);
+    cal.set(Calendar.SECOND, 0);
+    cal.set(Calendar.MILLISECOND, 0);
+    prevMinute = cal.getTime();
+
+    Job job = context.getHadoopJob();
+    if (!config.fileSystem.equals("S3")) {
+      LOG.warn("File system other than S3 not currently supported, using S3");
+    }
+    job.getConfiguration().set("fs.s3n.awsAccessKeyId", config.accessID);
+    job.getConfiguration().set("fs.s3n.awsSecretAccessKey", config.accessKey);
+
+    if (config.fileRegex != null) {
+      job.getConfiguration().set("input.path.regex", config.fileRegex);
+    }
+    job.getConfiguration().set("input.path.name", config.path);
+
+    if (config.timeTable != null) {
+      table = context.getDataset(config.timeTable);
+      String lastTimeRead = Bytes.toString(table.read("lastTimeRead"));
+      if (lastTimeRead == null) {
+        lastTimeRead = "0";
+      }
+      job.getConfiguration().set("last.time.read", lastTimeRead);
+    }
+
+    job.getConfiguration().set("cutoff.read.time", new SimpleDateFormat("yyyy-MM-dd-HH-mm").format(prevMinute));
+    if (config.inputFormatClass != null) {
+      Class<?> classType = Class.forName(config.inputFormatClass);
+      if (classType.isAssignableFrom(FileInputFormat.class)) {
+        job.setInputFormatClass((Class<? extends FileInputFormat>) classType);
+      } else {
+        LOG.warn("invalid input format class, using default TextInputFormat");
+      }
+    }
+    FileInputFormat.setInputPathFilter(job, BatchFileFilter.class);
+    FileInputFormat.addInputPath(job, new Path(config.path));
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Object> input, Emitter<StructuredRecord> emitter) throws Exception {
+    StructuredRecord output = StructuredRecord.builder(DEFAULT_SCHEMA)
+      .set("ts", System.currentTimeMillis())
+      .set("body", input.getValue().toString())
+      .build();
+    emitter.emit(output);
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    if (succeeded && table != null && "timefilter".equals(config.fileRegex)) {
+      table.write("lastTimeRead", new SimpleDateFormat("yyyy-MM-dd-HH-mm").format(prevMinute));
+    }
+  }
+
+  /**
+   * Config class that contains all the properties needed for the S3 source.
+   */
+  public static class S3BatchConfig extends PluginConfig {
+
+    @Name("fileSystem")
+    @Description(FILE_SYSTEM_DESCRIPTION)
+    private String fileSystem;
+
+    @Name("accessID")
+    @Nullable
+    @Description(ACCESS_ID_DESCRIPTION)
+    private String accessID;
```

Comment:
See my comment here for this and the next field - https://github.com/caskdata/cdap/pull/2997/files#r32881761